Identity factors

ccshan opened the pull request:
This is to fix #128

ccshan commented:
I've been working through `multi.py` and trying to make `EmbeddedTensor` support the `Tensor` operations used there. I'm looking at the `reshape` calls in `multi_mv` and `multi_solve`. Is it easy to explain how these are used? Is it just adding/removing dimensions of size 1? Does it ever even turn one dimension of size 6 into two dimensions of sizes 2 and 3, let alone something more nefarious? @davidweichiang

davidweichiang commented:
If there are nonterminals $X_1., .., X_n$, where $X_i$ has shape $(d_{i,1}, ..., d_{i,m_i})$, let $M = (\mathbb{R}^{d_{1,1}} \otimes \dots \otimes \mathbb{R}^{d_{1,m_1}}) \oplus \dots \oplus (\mathbb{R}^{d_{n,1}} \otimes \dots \otimes \mathbb{R}^{d_{n,m_n}})$. Then a MultiTensor is an element of $M$ or a linear transformation $M \rightarrow M$. The job of multi_mv(a,b) is to apply a to b, and the job of multi_solve is to solve the equation x = ax+b.

The purpose of the reshapes is to flatten one of the $\mathbb{R}^{d_{i,1}} \otimes \dots \otimes \mathbb{R}^{d_{i,m_i}}$ into a vector. So, it's not just squeezing/unsqueezing. Note that it's possible for $m_i = 0$, so this actually "flattens" a scalar into a 1-element vector.

ccshan commented:
Ok so if I understand you correctly, the granularity of each $d_{i,j}$ is that of a letter in an einsum equation, not finer but also not coarser. I should be able to preserve sparsity through a multi_mv operation...

davidweichiang commented:
Sorry I've been letting this sit so long. What would be needed to integrate this into the rest of the code? Are there particular things that you need me to check?

ccshan commented:
Hi! Aria is doing two things (digging a tunnel from both ends):
1. Write more tests for `einsum`
2. Replace `Tensor` by `EmbeddedTensor` in `MultiTensor` and see what breaks

Maybe you can make a quick start on #2? Maybe it will give rise to test failures that Aria can fix??

davidweichiang commented on fggs/semirings.py:
```diff
@@ -26,6 +27,10 @@ def zeros(self, shape: torch.Size) -> torch.Tensor:
         pass
 
+    @abstractmethod
+    def add_embedded(self, x: "EmbeddedTensor", y: "EmbeddedTensor") -> "EmbeddedTensor":
```
Is there any way to avoid adding all these `_embedded` methods?

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,801 @@
+    def logical_not(self) -> EmbeddedTensor:
+        return EmbeddedTensor(~self.physical, self.pembeds, self.vembeds, not self.default)
+
+    def log_softmax(self, dim) -> EmbeddedTensor:
```
sum_product.log_softmax has some extra checks for infinite inputs which are important. I think this drops those checks, right?

ccshan replied:
Just to clarify, the comment "make all the nonzero elements 1 and the zero elements 0" in sum_product.log_softmax means that the final result (due to log) is actually either 0=log(1) or -inf=log(0), right?

davidweichiang commented on fggs/semirings.py:
```diff
@@ -2,7 +2,8 @@
 from abc import ABC, abstractmethod
-from typing import Union
+from typing import Union, TYPE_CHECKING
+if TYPE_CHECKING: from fggs.indices import EmbeddedTensor
```
Is this to avoid a circular import? Maybe it's a sign that something needs to be refactored somehow.

ccshan commented:
Compared to the :heavy_check_mark: changes on this `identity` branch, I'm a lot less confident about the changes I made in the `multi_embedded` branch to get the sum_product tests to even run. But on that branch, we are down to just one test failure now:
```
======================================================================
ERROR: test_autograd (test.test_sum_product.TestSumProduct.test_autograd) (example='test/simplefgg.json (p=1.0)')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ccshan/u/rational/fmitf/fggs/test/test_sum_product.py", line 108, in test_autograd
    self.assertTrue(torch.autograd.gradcheck(f, in_values, atol=1e-3))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ccshan/u/rational/fmitf/python/lib/python3.11/site-packages/torch/autograd/gradcheck.py", line 1418, in gradcheck
    return _gradcheck_helper(**args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ccshan/u/rational/fmitf/python/lib/python3.11/site-packages/torch/autograd/gradcheck.py", line 1432, in _gradcheck_helper
    _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps,
  File "/home/ccshan/u/rational/fmitf/python/lib/python3.11/site-packages/torch/autograd/gradcheck.py", line 1075, in _gradcheck_real_imag
    gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
  File "/home/ccshan/u/rational/fmitf/python/lib/python3.11/site-packages/torch/autograd/gradcheck.py", line 1131, in _slow_gradcheck
    raise GradcheckError(_get_notallclose_msg(a, n, i, j, complex_indices, test_imag))
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 1,
numerical:tensor([[0.]], dtype=torch.float64)
analytical:tensor([[1.]], dtype=torch.float64)
```

davidweichiang commented on fggs/semirings.py:
```diff
@@ -91,17 +92,19 @@ class RealSemiring(Semiring):
-    
+
+    @staticmethod
+    def add(x, y):
```
Why no types on arguments?

ccshan replied:
Sorry, I didn't realize that overridden methods don't inherit types from the parent class. Fixing in https://github.com/diprism/fggs/pull/163/commits/5f87f66b10ea6cf13cb69831f02c0af28800fb2c

davidweichiang commented on fggs/semirings.py:
```diff
@@ -144,22 +148,25 @@ def from_int(self, n):
-                           x + torch.log1p(-torch.exp(d)), # type: ignore
-                           x + torch.log(-torch.expm1(d))).nan_to_num(nan=-inf, neginf=-inf, posinf=inf)
-    
+        d = y.sub(x)
```
Are the methods faster than the operators?

ccshan replied:
To work with both Tensor and EmbeddedTensors, the code must invoke methods (such as `t.exp()`) instead of functions (such as `torch.exp(t)`). But operators are methods (for instance, `y-x` means `y.__sub__(x)`).

Initially I was avoiding operator overloading because I was so unfamiliar with it. Now I think I can change `sub` to `__sub__` (and invoke it using operator syntax) if desired.

davidweichiang commented on fggs/sum_product.py:
```diff
@@ -99,10 +101,23 @@ def F(fgg: FGG, x: MultiTensor, inputs: MultiTensor, semiring: Semiring) -> Mult
     return Fx
 
 
+def print_duplicate(loc: str, xs: Sequence[Node], ys: Sequence[Node]) -> bool:
```
Is this just for debugging? Does it need to stay?

ccshan replied:
It's just for debugging. It doesn't need to stay. I just feel like it might be useful for future debugging...?

davidweichiang commented on fggs/sum_product.py:
```diff
@@ -289,12 +317,13 @@ class SumProduct(torch.autograd.Function):
     def forward(ctx, fgg: FGG, opts: Dict, in_labels: Sequence[EdgeLabel], out_labels: Sequence[EdgeLabel], *in_values: Tensor) -> Tuple[Tensor, ...]: # type: ignore
         ctx.fgg = fgg
         method, semiring = opts['method'], opts['semiring']
+        from_dense = lambda t: EmbeddedTensor(t, default=semiring.from_int(0).item())
```
Could also write this as a nested `def`

ccshan replied:
Oh yeah done thanks

ccshan commented on fggs/sum_product.py:
```diff
@@ -205,38 +232,30 @@ def sum_product_edges(fgg: FGG, nodes: Iterable[Node], edges: Iterable[Edge], ex
         eshape = fgg.shape(ext)
         vshape = [s if n in connected else 1 for n, s in zip(ext, eshape)]
-        rshape = [1 if n in connected else s for n, s in zip(ext, eshape)]
-        out = out.view(*vshape).repeat(*rshape) # TODO: why not use expand rather than repeat?
```
This code now uses `EmbeddedTensor.expand`

ccshan commented on fggs/semirings.py:
```diff
@@ -119,6 +122,7 @@ def callback(compute_sum):
             return compute_sum(torch_semiring_einsum.utils.add_in_place,
                                torch_semiring_einsum.utils.sum_block,
                                multiply_in_place)
+        # TODO: Why blocksize=1?
```
I just tried using AutomaticBlockSize everywhere. It works but requires https://github.com/bdusell/semiring-einsum/pull/34

davidweichiang commented:
On a larger scale, I'm still wondering about the relationship of Tensor, Semiring, and EmbeddedTensor. Do we have it right? The fact that Semiring uses Tensors seems backwards, but it had to be that way because of the difficulty of subclassing Tensor to use other semirings. OTOH, EmbeddedTensor already goes through the trouble of (sort of) subclassing Tensor, which makes me wonder whether the relationship between EmbeddedTensor and Semiring could be put right. That is, maybe an EmbeddedTensor could contain a Semiring and a Tensor, so that when you, e.g., add two EmbeddedTensors, it looks up the addition operation from the Semiring.

davidweichiang reviewed (approved):
I checked all the other files besides indices.py, and I skimmed through indices.py but can't say I understand all of it.

Is the complexity of EmbeddedTensor going to be an obstacle to a user who wants to use FGGs but not PERPL? Will it be possible to make their use optional (possibly contrary to my previous comment about the relationship with Semiring and Tensor)?

ccshan commented:
Currently `EmbeddedTensor(t)` converts `t:Tensor` into an `EmbeddedTensor`, and `et.to_dense()` converts `et:EmbeddedTensor` into a `Tensor`. We can try to make these conversions more implicit, but what kind of "using FGGs but not PERPL" are we trying to ease? Currently, the `sum_product` and `viterbi` code only uses `EmbeddedTensor` internally (but in order to support product and sum types, I'm working to change that), so it's a good time to think about what the interface should be (should `sum_product` sometimes return an `EmbeddedTensor`?). It might make sense to make `sum_product` and `viterbi` operate on an abstract `TensorLike` type, like some `Semiring` methods do.

One reason to not put a Semiring inside (Embedded)Tensors is the binary method problem: what should `t1+t2` do if `t1` and `t2` contain different Semirings?

If you could point out particular places in `indices.py` that are hard to understand, I'm sure there are plenty of comments to add...

davidweichiang commented:
Since the word "embedding" is already overused in a ML context, are there any other names you can think of? StructuredTensor?

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+"""
+# Algebraic index types
+
+A tensor dimension often corresponds to a random variable, and indices along
```
I try to avoid the word "dimension" because it has two meanings: a 3x3 matrix, considered as a multidimensional array, has 2 dimensions, but in a linear algebra context, it is a linear mapping on a vector space with 3 dimensions (or could itself be thought of as a vector with 9 dimensions!)

ccshan replied:
I totally agree so what **is** the word that people have settled on for the thing that's counted by `ndim`?? PyTorch documentation says "dimension" :-/

ccshan replied:
Ah you use the word "axis"

davidweichiang replied:
NumPy inconsistently uses both "axis" and "dimension"; PyTorch uses "dimension." Other terms are "index", "mode", or "way". For the named tensor notation paper, we used "axis".

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+to build up three embeddings such as [X(5) * Y(7), X(5), Y(7)].  The three
+embeddings determine the shape of the virtual tensor to be [35, 5, 7].  Element
+(i,j) of the matrix corresponds to element (i*7+j,i,j) of the tensor.  Other
+elements of the tensor are presumed to be zero.
```
It might be helpful to choose smaller numbers and actually write the index embedding out. Or actually, I feel that writing the inverse out would be even more helpful, like for the example below,

```
[ [ 0,    0,    0, ... ],
  [ m[0], 0,    0, ...],
  [ 0,    m[1], 0, ...],
  ... ]
```

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+    """An injective mapping from physical indices to virtual indices."""
+
+    @abstractmethod
+    def depict(self, letterer: Callable[[EmbeddingVar], str]) -> str:
```
please add description

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+        warn(f"Attempt to unify {e.depict(debugging_letterer)} and {f.depict(debugging_letterer)} indicates index type mismatch")
+        return False
+
+    def antiunify(e: Embedding, f: Embedding, antisubst: AntiSubst) -> Embedding:
```
I'm not familiar with antiunification and am guessing a lot of other people also aren't

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+
+    @abstractmethod
+    def numel(self) -> int:
+        """The virtual indices in the image of this mapping lie in [0,numel())."""
```
This comment doesn't seem right, since virtual indices are in general tuples?

davidweichiang replied:
Please ignore the above, but if you can think of a way to name/document to make it clear that an EmbeddedTensor has multiple Embeddings, that could help. For example an Embedding could instead be called a VirtualAxis or something.

ccshan replied:
If `EmbeddingVar` and `ProductEmbedding` and `SumEmbedding` becomes `PhysicalAxis` and `ProductAxis` and `SumAxis`, then should `Embedding` become `Axis` or `VirtualAxis`?

ccshan commented:
> Since the word "embedding" is already overused in a ML context, are there any other names you can think of? StructuredTensor?

How about "pattern"?
* "embedding" -> "(sparsity) pattern"
* "embedding var(iable)" -> "(sparsity-)pattern var(iable)"
* "embedded tensor" -> "(sparsity-)patterned tensor"

hydropyrum commented:
That works for me.

davidweichiang commented:
That works for me.

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+    def dim(self) -> int:
+        return len(self.vembeds)
+
+    ndim = ndimension = property(dim)
```
Is ndimension ever used?

ccshan replied:
It's not. I just put it in because I saw it next to ndim in the pytorch documentation. But oh wait, I got it wrong, `ndimension` is a method like `dim`, not a property like `ndim`! I'll fix that.

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+        """Express this embedding as a product of non-product embeddings."""
+        yield self
+
+    def forward(self, subst: Subst) -> Embedding:
```
Confusable with the forward method of torch.nn.Function

davidweichiang commented on fggs/indices.py:
```diff
@@ -0,0 +1,1293 @@
+
+    @abstractmethod
+    def stride(self, subst: Subst) -> Tuple[int, Dict[EmbeddingVar, int]]:
+        """The coefficients of the affine map from physical to virtual indices."""
```
Please add (if correct): if (offset, strides) = self.stride(), then the physical indices {x1: i1, x2:i2, ...}  map to virtual index offset+strides[x1]+strides[x2]+....

davidweichiang commented:
Is it easy to explain why EmbeddingVar isn't just an int (0 for the first physical axis, 1 for the second physical axis, etc.)? And what Substs are used for?

ccshan commented:
> Is it easy to explain why EmbeddingVar isn't just an int (0 for the first physical axis, 1 for the second physical axis, etc.)?

Is it enough to say that we want to distinguish between the first physical axis of one PatternedTensor and the first physical axis of another PatternTensor?

> And what Substs are used for?

Unification is first and foremost used to implement einsum. For example to einsum "ij,jk->ik" (in other words matrix multiplication), we unify the second virtual axis of the first PatternedTensor against the first virtual axis of the second PatternedTensor, and that might add some entry to the Subst that maps some physical axis of the first PatternedTensor to some product of physical axes of the second PatternedTensor.

ccshan commented:
Hmm, why are my last 7 commits to the identity branch (ending with 89e838d) not shown in the pull request? Anyway, I think I have addressed all your comments.
